upgrade/upgrade.go

```diff
@@ -148,61 +148,55 @@ func doAPIVersionUpgrade(namespace string, config *restclient.Config, lhClient *
 
 	crdAPIVersion := ""
 
-	crdAPIVersionSetting, err := lhClient.LonghornV1beta1().Settings(namespace).Get(context.TODO(), string(types.SettingNameCRDAPIVersion), metav1.GetOptions{})
+	crdAPIVersionSetting, err := lhClient.LonghornV1beta2().Settings(namespace).Get(context.TODO(), string(types.SettingNameCRDAPIVersion), metav1.GetOptions{})
 	if err != nil {
 		if !apierrors.IsNotFound(err) {
 			return err
 		}
-		// v1alpha1 doesn't have a setting entry for crd-api-version, need to create
-		crdAPIVersionSetting = &longhorn.Setting{
-			ObjectMeta: metav1.ObjectMeta{
-				Name: string(types.SettingNameCRDAPIVersion),
-			},
-			Value: "",
-		}
-		crdAPIVersionSetting, err = lhClient.LonghornV1beta1().Settings(namespace).Create(context.TODO(), crdAPIVersionSetting, metav1.CreateOptions{})
-		if err != nil {
-			return errors.Wrap(err, "cannot create CRDAPIVersionSetting")
-		}
 	} else {
 		crdAPIVersion = crdAPIVersionSetting.Value
 	}
 
 	if crdAPIVersion != "" &&
-		crdAPIVersion != types.CRDAPIVersionV1alpha1 &&
-		crdAPIVersion != types.CRDAPIVersionV1beta1 {
+		crdAPIVersion != types.CRDAPIVersionV1beta1 &&
+		crdAPIVersion != types.CRDAPIVersionV1beta2 {
 		return fmt.Errorf("unrecognized CRD API version %v", crdAPIVersion)
 	}
+
 	if crdAPIVersion == types.CurrentCRDAPIVersion {
 		logrus.Info("No API version upgrade is needed")
 		return nil
 	}
 
 	switch crdAPIVersion {
 	case "":
-		// it can be a new installation or v1alpha1
-		isV1alpha1, err := v1alpha1.IsCRDVersionMatch(config, namespace)
+		// upgradable: new installation
+		// non-upgradable: error or non-supported version (v1alpha1 which cannot upgrade directly)
+		upgradable, err := v1beta1.CanUpgrade(config, namespace)
 		if err != nil {
-			logrus.Warnf("Cannot verify current CRD version, assume it's not v1alpha1: %v", err)
+			return err
 		}
-		if !isV1alpha1 {
-			crdAPIVersionSetting.Value = types.CurrentCRDAPIVersion
-			if _, err := lhClient.LonghornV1beta1().Settings(namespace).Update(context.TODO(), crdAPIVersionSetting, metav1.UpdateOptions{}); err != nil {
-				return errors.Wrapf(err, "cannot finish CRD API upgrade by setting the CRDAPIVersionSetting to %v", types.CurrentCRDAPIVersion)
+
+		if upgradable {
+			crdAPIVersionSetting = &longhorn.Setting{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: string(types.SettingNameCRDAPIVersion),
+				},
+				Value: types.CurrentCRDAPIVersion,
 			}
-			logrus.Infof("Initialized CRD API Version to %v", types.CurrentCRDAPIVersion)
-			break
-		}
-		fallthrough
-	case types.CRDAPIVersionV1alpha1:
-		if types.CurrentCRDAPIVersion != types.CRDAPIVersionV1beta1 {
-			return fmt.Errorf("cannot upgrade from %v to %v directly", types.CRDAPIVersionV1alpha1, types.CurrentCRDAPIVersion)
+			_, err = lhClient.LonghornV1beta2().Settings(namespace).Create(context.TODO(), crdAPIVersionSetting, metav1.CreateOptions{})
+			if err != nil {
```

Comment:
Consider adding a check for `!apierrors.IsAlreadyExists(err)`.

This could prevent a corner case that the CR `crd-api-version` is there but with an empty string, then in longhorn-manager upgrade path, the longhorn-manager would try to create the `crd-api-version` Setting CR and return an error since the CR already exists.